Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,446 @@
+            case removeConnection
+            case keepConnection
+        }
+        
+        /// Cleanup the current connection for shutdown.
+        ///
+        /// This method is called, when the connections shall shutdown. Depending on the state
+        /// the connection is in, it adds itself to one of the arrays that are used to signal shutdown
+        /// intent to the underlying connections. Connections that are backing off can be easily
+        /// dropped (since, we only need to cancel the backoff timer), connections that are leased
+        /// need to be cancelled (notifying the `ChannelHandler` that we want to cancel the
+        /// running request), connections that are idle can be closed right away. Sadly we can't
+        /// cancel connection starts right now. For this reason we need to wait for them to succeed
+        /// or fail until we finalize the shutdown.
+        ///
+        /// - Parameter context: A cleanup context to add the connection to based on its state.
+        /// - Returns: A cleanup action indicating if the connection can be removed from the
+        ///            connection list.
+        func cleanup(_ context: inout CleanupContext) -> CleanupAction {
+            switch self.state {
+            case .starting:
+                return .keepConnection
+            case .backingOff:
+                context.connectBackoff.append(self.connectionID)
+                return .removeConnection
+            case .active(let connection, _, let usedStreams, _):
+                if usedStreams <= 0 {
+                    context.close.append(connection)
+                    return .removeConnection
+                } else {
+                    context.cancel.append(connection)
+                    return .keepConnection
+                }
+            case .draining(let connection, _, _):
+                context.cancel.append(connection)
+                return .keepConnection
+            case .closed:
+                preconditionFailure("Unexpected state for cleanup: Did not expect to have connections with this state in the state machine: \(self.state)")
+            }
+        }
+
+        init(connectionID: Connection.ID, eventLoop: EventLoop) {
+            self.connectionID = connectionID
+            self.eventLoop = eventLoop
+            self.state = .starting
+        }
+    }
+    
+    struct HTTP2Connections {
+        /// A connection's use. Did it serve in the pool or was it specialized for an `EventLoop`?
+        enum ConnectionUse {
+            case generalPurpose
+            case eventLoop(EventLoop)
+        }
+        
+        /// Information around an idle connection.
+        struct IdleConnectionContext {
+            /// The `EventLoop` the connection runs on.
+            var eventLoop: EventLoop
+            /// The connection's use. Either general purpose or for requests with `EventLoop`
+            /// requirements.
+            var use: ConnectionUse
+        }
```

Comment:
We don't need this buddy, I think. we have the better fitting `AvailableConnectionContext`.